[Android] Use headingLevel for heading accessibility property

mnayef95 opened the pull request:
[![talabat.com](https://img.shields.io/badge/talabat.com-contributions-FF5A00?style=flat&logo=flutter&logoColor=white)](https://www.talabat.com) [![Talabat Flutter PRs](https://img.shields.io/badge/Talabat_Flutter_PRs-10%20merged-97ca00?style=flat&logo=flutter&logoColor=white)](https://github.com/search?q=org%3Aflutter+talabat&type=pullrequests)

Fixes #174150

This PR updates the Android `AccessibilityBridge` to use the `headingLevel` property from `SemanticsNode` to determine if a node should be marked as a heading for accessibility purposes.

Previously, the `IS_HEADER` flag was used. Now, a node is considered a heading if `headingLevel > 0`. This aligns with the common accessibility pattern where heading levels (like H1, H2, etc.) define the structure.

The following changes were made:
- Modified `AccessibilityBridge.java` to set the heading status based on `semanticsNode.headingLevel > 0` for Android API level 28 and above.
- Added `headingLevel` to the `SemanticsNode` data structure in Java and its serialization/deserialization logic.
- Updated the C++ `PlatformViewAndroidDelegate` to include `headingLevel` when serializing `SemanticsNode` data.
- Increased `kBytesPerNode` in `platform_view_android_delegate.h` to account for the new `headingLevel` field.
- Added tests in `AccessibilityBridgeTest.java` to verify the new heading logic:
    - Nodes with `headingLevel = 0` are not headings.
    - Nodes with `headingLevel > 0` are headings.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.


<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

mnayef95 committed:
Replace IS_HEADER SemanticsFlag with headingLevel

This commit introduces a `headingLevel` property to SemanticsNode and updates the Android `AccessibilityBridge` to use this new property instead of the `IS_HEADER` flag.

- A `headingLevel` of 0 indicates that the node is not a heading.
- Any positive `headingLevel` indicates that the node is a heading of that level.
- The `IS_HEADER` SemanticsFlag has been removed.
- Android's `AccessibilityNodeInfo.isHeading` is now set based on whether `headingLevel > 0`.
- Tests have been added to verify the new heading level behavior.
- The size of `SemanticsNode` in `PlatformViewAndroidDelegate` has been updated to accommodate the new `headingLevel` field.

mnayef95 committed:
Add heading level to accessibility node data

This change adds a `heading_level` field to the `FlutterSemanticsNode` and related structures. This allows assistive technologies to announce the heading level of elements in the UI.

The `heading_level` is an integer where 0 indicates the node is not a heading, and higher values (1, 2, ...) represent the heading rank.

This is implemented by:
 - Adding `heading_level` to `FlutterSemanticsNode` in `embedder.h`.
 - Adding `heading_level` to `SemanticsNode` in `accessibility_bridge.h`.
 - Populating `result.heading_level` from `flutter_node.heading_level` in `AccessibilityBridge::GetFlutterPlatformNodeDelegateFromFlutterNode`.
 - Initializing `heading_level` to 0 in unit tests for `FlutterPlatformNodeDelegate`.

chinmaygarde commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -1565,6 +1565,10 @@ typedef struct {
+  /// The heading level for this node. A value of 0 means the node is not a
+  /// heading; higher values (1, 2, …) indicate the heading rank, with lower
+  /// numbers being higher-level headings.
+  int32_t heading_level;
```
The addition of the member here and another in the `FlutterSemanticsNode2` struct constitutes a breaking ABI change. To avoid this, these members needs to be added at the bottom of the struct. To learn more about rationale, please see the comment on the "Application Binary Interface" detailed at the top of this header.

The change to the cross-platform stable API also warrants a test. See `embedder_a11y_unittests.cc` for a a spot. Adding a new assert to an existing test is likely sufficient.

mnayef95 replied:
@chinmaygarde I was able to fix it by changing the test from: `FlutterSemanticsNode2 root;` to: `FlutterSemanticsNode2 root{sizeof(FlutterSemanticsNode2), 0};`

<img width="854" height="617" alt="Screenshot 2025-09-19 at 2 06 43 PM" src="https://github.com/user-attachments/assets/efc88e61-f4b1-4270-920f-36f15037cd9e" />

mnayef95 replied:
Hello @chinmaygarde, I would appreciate another review, thank you

chinmaygarde reviewed (dismissed):
The major concern is the breaking ABI change. Left some suggestions on how to avoid it.

mnayef95 committed:
Refactor: Move heading_level to end of FlutterSemanticsNode struct

The `heading_level` member of the `FlutterSemanticsNode` and `FlutterSemanticsNode2` structs was moved to the end of the struct. This change also updates the corresponding frozen struct and unit tests.

mnayef95 committed:
Fix order of fields in SemanticsNode struct

The order of `heading_level` and `child_count` in the `SemanticsNode` struct initialization in `flutter_platform_node_delegate_unittests.cc` was inconsistent with the struct definition in `accessibility_bridge.h`.

This change reorders the initialization in the unit test to match the struct definition.

mnayef95 committed:
Add test for heading_level

mnayef95 committed:
Fix TestAccessibilityBridge in flutter_platform_node_delegate_unittests.cc

The `FlutterSemanticsNode2` struct was not being initialized correctly in the `TestAccessibilityBridge`. This change initializes the `struct_size` and `id` fields of the struct.

mnayef95 committed:
[Test] Remove redundant ID assignments in FlutterSemanticsNode2

The `id` field of `FlutterSemanticsNode2` is already initialized by the constructor when providing it as the second argument. This change removes the redundant explicit assignments of the `id` field after object creation in the unit tests.

mnayef95 commented:
Hello @ash2moon I would appreciate your review here

chunhtai commented on engine/src/flutter/shell/platform/android/io/flutter/view/AccessibilityBridge.java:
```diff
@@ -2197,7 +2197,6 @@ public enum Action {
     HAS_ENABLED_STATE(1 << 6),
     IS_ENABLED(1 << 7),
     IS_IN_MUTUALLY_EXCLUSIVE_GROUP(1 << 8),
-    IS_HEADER(1 << 9),
```
Unless lint is complaining this, it will be better if we keep this for readability. but I am not android expert, not sure if there is a real consequence of keeping unused define

ash2moon replied:
I agree, we should keep it for readability. There should be no issues of keeping this defined and unused.

chunhtai reviewed (approved):
logic wise LGTM, leaving it to Android team to do a final review

ash2moon reviewed (commented):
Looks good from android side!

mnayef95 committed:
Add IS_HEADER flag to accessibility states in AccessibilityBridge
